contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,276 @@
+
+  - **what** happened?
+  - **when** did it happen?
+  - **who** initiated it?
+  - **on what** did it happen (e.g. pod foo/bar)?
+  - **where** was it observed (e.g. apiserver hostname)?
+  - from **where** was it initiated? (e.g. kubectl IP)
+  - to **where** was it going? (e.g. node 1.2.3.4 for kubectl proxy).
+
+1. Depending on the environment, as a cluster operator I want to **define the amount of audit logging**, balancing computational overhead for the apiserver with the detail and completeness of the log.
+
+1. As a cluster operator I want to **integrate with external systems**, which will have different requirements for the log format, network protocols and communication modes (e.g. pull vs. push).
+
+1. As a cluster operator I must be able to provide a **complete trace of changes to an object** to API objects.
+
+1. As a cluster operator I must be able to create a trace for **all accesses to a secret**.
+
+1. As a cluster operator I want to **define log-rotation** of a file-based output backend.
+
+1. As a cluster operator I must be able to log non-CRUD access like **kubectl exec**, when it started, when it finished and with which initial parameters.
+
+### Out of scope use-cases
+
+1. As a cluster operator I must be able to get a trace of interactive commands executed in **kubectl exec**, **attach** or **run**.
+
+## Community Work
+
+- Kubernetes basic audit log PR: https://github.com/kubernetes/kubernetes/pull/27087/
+- OpenStack's implementation of the CADF standard: https://www.dmtf.org/sites/default/files/standards/documents/DSP2038_1.1.0.pdf
+- Cloud Auditing Data Federation standard: https://www.dmtf.org/standards/cadf
+- Ceilometer audit blueprint: https://wiki.openstack.org/wiki/Ceilometer/blueprints/support-standard-audit-formats
+- Talk from IBM: An Introduction to DMTF Cloud Auditing using
+the CADF Event Model and Taxonomies https://wiki.openstack.org/w/images/e/e1/Introduction_to_Cloud_Auditing_using_CADF_Event_Model_and_Taxonomy_2013-10-22.pdf
+
+## Architecture
+
+When implementing audit logging there are basically two options:
+
+1. put a logging proxy in front of the apiserver
+2. integrate audit logging into the apiserver itself
+
+Both approaches have advantages and disadvanteges:
+
+- **pro proxy**:
+  + keeps complexity out of the apiserver
+  + reuses existing solutions
+- **contra proxy**:
+  + has no deeper insight into the Kubernetes api
+  + has no knowledge of auth, authn, admission
+  + has no access to the storage level for diffential output
+  + has to terminate SSL and complicates client certificates based auth
+
+In the following the second approach is described without a proxy.
+
+## Proposed Design
+
+The main concepts are those of
+
+- an audit *event*,
+- an audit *policy*,
+- an audit *policy action*,
+- an audit *output backend*.
```

Comment:
Not quite, because the shape of the output will be entirely up to audit mechanics (be that JSON, txt, etc) the backed is only responsible for storing that data. What happens further with it is out of scope of this doc.